A new algorithm for polygon clipping in solar calculations for rectangular surfaces

xuanluo113 opened the pull request:
Pull request overview
---------------------
There exists an extensive amount of literature on polygon clipping when constrained to rectangular surfaces. The current polygon clipping method implemented in EnergyPlus, using the *Sutherland-Hodgman* algorithm, is a performance hotspot. The method is generalized for clipping area of all shapes, but in EnergyPlus simulations runs, are often called with only rectangular surfaces.

We propose a negligible amount of code to check if the clipping surface is rectangular. And if so,
we redirect to a new method using the `Slater & Barsky` algorithm.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

xuanluo113 committed:
NFP and test code

xuanluo113 committed:
fix

xuanluo113 committed:
idd

mjwitte commented:
@XuanLuoLBL This needs a unit test, please. Once that's in, please request a reviewer here in the PR.

mjwitte commented:
@XuanLuoLBL Needs doc updates, also.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into shadow-clip

xuanluo113 committed:
unit test

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into shadow-clip

xuanluo113 committed:
docs

xuanluo113 commented:
> @XuanLuoLBL This needs a unit test, please. Once that's in, please request a reviewer here in the PR.

@mjwitte This branch is also ready for review, as I added those unit tests and documentation.

xuanluo113 commented:
@Myoldmopar Edwin, this branch is ready for review. Would you please suggest reviewers? Thanks.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into shadow-clip

mjwitte committed:
Fix OutputFiles stuff in new polygon unit test

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
```diff
@@ -291,6 +293,53 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
 
 The Sutherland-Hodgman algorithm (Sutherland, Hodgman, 1974) is less complex compared to the Weiler-Atherton method and is well-suited to clipping convex polygons. In actuality, only convex shading surfaces are currently supported by EnergyPlus. Let X be a polygon called the ``subject polygon'' (SP) and Y be a polygon called the ``clipping polygon'' (CP). The method performs the computation by it ...
 
+The Slater-Barsky algorithm can also be chosen for clipping surfaces which are rectangular. The algorithm further saves computational cost. It relies on converting the line segment to be clipped into a parametric equation. Below is the general form for the x- and y- components of a parametric line:
```
Is there a reference for this method?

xuanluo113 replied:
Reference added

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
```diff
@@ -277,6 +277,8 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
   Convex Weiler - Atherton
 \item
   Sutherland -- Hodgman
+\item
```
Just before this "Two methods for polygon clipping" should be "Three".

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
```diff
@@ -277,6 +277,8 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
 \item
   Sutherland -- Hodgman
+\item
+  Slater -- Barsky (for rectangular surfaces only)
```
All of these should simply be Weiler-Atherton, Sutherland-Hodgman, Slater-Barsky, single dash and no spaces. The others were probably messed up in the original latex translation.

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
```diff
@@ -291,6 +293,53 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
+
+Distances from the endpoints to the edges of the clipping rectangle that collide with the line are used to obtain two values of $t$ that parameterize the equation, representing the pair of new endpoints.
+
+Once we have the two values of $t, t_1 and t_2$, we can calculate the clipped line endpoints (Note that $t_2 > t_1$).
```
Not sure why, but this shows in the pdf as *t, t1 andt2*. Try closing and reopening math mode around the and. `$t, t_1$ and $t_2$`.

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
```diff
@@ -291,6 +293,53 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
+y'_2 = y_0 + t_2 \Delta y
+\end{equation}
+
+The "Slater \& Barsky" algorithm uses space subdivision to reduce the number of pre-emptive calculations. We break the plane into 9 parts, where region (4) is the clipping region.
```
Replace this with simply Slater-Barsky (no quotes).

mjwitte commented on doc/engineering-reference/src/climate-sky-and-solar-shading-calculations/shading-module.tex:
````diff
@@ -291,6 +293,53 @@ \subsection{Polygon Clipping Algorithms}\label{polygon-clipping-algorithms}
+| 6 | 7 | 8 |
+| 3 | 4 | 5 |
+| 0 | 1 | 2 |
+```
````
This is rendering all on one line in the pdf: “‘ | 6 | 7 | 8 | | 3 | 4 | 5 | | 0 | 1 | 2 | “‘
Looks like you were trying for format this as code?

xuanluo113 replied:
Using a line break instead.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -558,6 +558,7 @@ ShadowCalculation,
        \type choice
        \key ConvexWeilerAtherton
        \key SutherlandHodgman
+       \key SlaterBarskyandSutherlHodgman
```
This should be SlaterBarskyandSutherl*and*Hodgman or SlaterBarksyAndSutherlanHodgman

mjwitte commented on src/EnergyPlus/DataSystemVariables.cc:
```diff
@@ -118,6 +118,7 @@ namespace DataSystemVariables {
     std::string const cIgnoreBeamRadiation("IgnoreBeamRadiation");
     std::string const cIgnoreDiffuseRadiation("IgnoreDiffuseRadiation");
     std::string const cSutherlandHodgman("SutherlandHodgman");
+    std::string const cSlaterandBarsky("SlaterandBarsky");
```
Nit picking, for consistency, this should be "SlaterBarsky" here and everywhere else.

mjwitte commented on src/EnergyPlus/SimulationManager.cc:
```diff
@@ -1765,7 +1765,11 @@ namespace SimulationManager {
         if (SutherlandHodgman) {
-            ObjexxFCL::gio::write(EchoInputFile, fmtLD) << "ClippingAlgorithm=SutherlandHodgman";
+            if (SlaterandBarsky) {
+                ObjexxFCL::gio::write(EchoInputFile, fmtLD) << "ClippingAlgorithm=SlaterBarskyandSutherlHodgman";
```
SlaterBarskyandSutherlandHodgman

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -3819,6 +4218,7 @@ namespace SolarShading {
         // Using/Aliasing
         using General::ReallocateRealArray;
         using General::SafeDivide;
+        using DataSystemVariables::SlaterandBarsky;
```
No new `using` statements, please. Just use the full reference below.

mjwitte commented on tst/EnergyPlus/unit/SolarShading.unit.cc:
```diff
@@ -1462,3 +1464,384 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_DisableGroupSelfShading)
+
+TEST_F(EnergyPlusFixture, SolarShadingTest_PolygonClippingDirect)
+{
+    std::string const idf_objects = delimited_string({"  Version,9.1;                                                                       ",
```
Version object isn't necessary here.

mjwitte commented on tst/EnergyPlus/unit/SolarShading.unit.cc:
```diff
@@ -1462,3 +1464,384 @@ TEST_F(EnergyPlusFixture, SolarShadingTest_DisableGroupSelfShading)
+                                                      "  SurfaceConvectionAlgorithm:Outside,TARP;                                           ",
+                                                      "  HeatBalanceAlgorithm,ConductionTransferFunction;                                   ",
+                                                      "  Timestep,6;                                                                        ",
+                                                      "  RunPeriod,                                                                         ",
```
Minor, but I don't think SurfaceConvectionAlgorithm:*, HeatBalanceAlgorithm, and Timestep are needed here. Strive to keep unit test idf snippets as small as possible. Seems the constructions could be simplified for this test.

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -3790,13 +3808,394 @@ namespace SolarShading {
+
+    inline void CLIPLINE(Real64 &x1, Real64 &x2, Real64 &y1, Real64 &y2,
+                         Real64 maxX, Real64 minX, Real64 maxY, Real64 minY, bool &visible, bool &rev)  {
+        // SLATER & BARSKY 1994
```
A more complete citation would be good here. And the new functions should have at least a couple of comment lines stating what they do.

Myoldmopar replied:
Yeah, this looks fine, but the citation is important.  The CLIPLINE and CLIPRECT function names look suspiciously like they were taken from some FORTRAN source.  Can you verify where this exact code came from?  Also I'm not sure you're going to gain anything by trying to `inline` this massive function filled with logic.

xuanluo113 replied:
@Myoldmopar @mjwitte I added the references and am trying to add line comments. I don't think it's borrowed code. It should be based on the algorithm introduced in the paper mimicking the Fortran naming style of other functions.

mjwitte commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -3867,6 +4267,21 @@ namespace SolarShading {
+                                              HCX[l1 + 2] == HCX[l1 + 3]) ||
+                                             ((((HCY[l1] == HCY[l1 + 1] && HCX[l1] != HCX[l1 + 1]) &&
+                                                (HCX[l1 + 2] == HCX[l1 + 1] && HCX[l1 + 3] == HCX[l1])) &&
+                                               (HCY[l1 + 2] == HCY[l1 + 3]))))));
```
This is hard to read. Apply formatting? Also, seems it would be more efficient to move `(NV2==4)` to an if block wrapping this so there's no need to check the other conditions if `NV2`<>4.

Myoldmopar replied:
As an FYI, native C++ operators will short circuit, so if NV2 is not 4, it will not evaluate the rest of this.  But that said, yeah, this is a big messy expression.  I would create three individual terms:

```c
bool const numVertsIs4 = NV2 == 4;
bool const term1 = blah;
bool const term2 = blah;
rectFlat = numVertsIs4 || term1 || term2;
```

Of course better naming but that would at least make the logic here tractable.

mjwitte commented on doc/input-output-reference/src/overview/group-simulation-parameters.tex:
```diff
@@ -562,7 +562,7 @@ \subsubsection{Inputs}\label{inputs-10-019}
 \paragraph{Field: Polygon Clipping Algorithm}\label{field-polygon-clipping-algorithm}
 
-This is an advanced feature. Prior to V7, the internal polygon clipping method was a special case of the Weiler-Atherton method. Now, two options are available: \textbf{SutherlandHodgman} (default) and \textbf{ConvexWeilerAtherton}. Theoretically, Sutherland-Hodgman is a simpler algorithm but it works well in cases where receiving surfaces (of shadows) are non-convex. The Weiler-Atherton implemen ...
+This is an advanced feature. Prior to V7, the internal polygon clipping method was a special case of the Weiler-Atherton method. Now, three options are available: \textbf{SutherlandHodgman} (default), \textbf{SlaterBarskyandSutherlHodgman} and \textbf{ConvexWeilerAtherton}. Theoretically, Sutherland-Hodgman is a simpler algorithm but it works well in cases where receiving surfaces (of shadows) ar ...
```
The formatting of the option names here is inconsistent. The pdf renders some in bold, some in single quotes, some in *'s (mixing md and latex?). This would be much better using the `{description}` method, example [here](https://github.com/NREL/EnergyPlus/blob/55f5f74c4f7b5b99e0110e5c7a158fd85bb20cc2/doc/input-output-reference/src/overview/group-surface-construction-elements.tex#L4083-L4088).

mjwitte reviewed (commented):
@xuanluo113 I've reviewed the idd and documentation changes. Comments below.
@Myoldmopar I've done a cursory review of the new code. Would like your eyes on this also.

mjwitte commented:
@xuanluo113 Be sure to pull this branch before making additional changes. I merged in develop yesterday and fixed the unit tests for the new eio file stuff.

Myoldmopar commented on src/EnergyPlus/SolarShading.cc:
```diff
@@ -657,25 +658,42 @@ namespace SolarShading {
             } else if (UtilityRoutines::SameString(cAlphaArgs(2), "ConvexWeilerAtherton")) {
                 SutherlandHodgman = false;
                 cAlphaArgs(2) = "ConvexWeilerAtherton";
+            } else if (UtilityRoutines::SameString(cAlphaArgs(2), "SlaterBarskyandSutherlHodgman")) {
```
Is there a reason that `Sutherland` was shortened in this choice field?

Myoldmopar reviewed (commented):
I made a few comments that should at least be discussed, and maybe requires some code changes.

xuanluo113 committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into shadow-clip

xuanluo113 committed:
address comments - IDD

xuanluo113 commented:
@Myoldmopar @mjwitte I've addressed all comments except for the coding style and comment ones for the two newly added solar clipping functions. Still working on this.

Myoldmopar commented:
Regressions locally look all clean, just eio diffs.  I'm merging this.

xuanluo113 commented:
> Regressions locally look all clean, just eio diffs. I'm merging this.

thanks.
